src/java.base/share/classes/jdk/internal/math/FormattedFPDecimal.java

```diff
@@ -69,6 +68,84 @@ public static FormattedFPDecimal valueOf(double v, int prec, char form) {
         };
     }
 
+    private static FormattedFPDecimal split(double v) {
+        FormattedFPDecimal fd = new FormattedFPDecimal();
+        DoubleToDecimal.split(v, fd);
+        return fd;
+    }
+
+    /**
+     * Returns a FormattedFPDecimal with the appropriate precision for
+     * {@link Double#toString(double)}.
+     *
+     * @see java.math.BigDecimal#valueOf(double)
+     */
+    public static FormattedFPDecimal valueForDoubleToString(double v) {
+        FormattedFPDecimal fd = split(v);
+        final int expR = fd.getExponentRounded();
+
+        // Adjust precision, following rules for Double.toString. There is
+        // always at least one digit and some cases require an extra one to
+        // force a digit after the decimal. No additional rounding is performed;
+        // no significant trailing digits are removed.
+
+        final int targetPrec = switch (expR) {
+            case -3, -2, -1 ->
+                // No extra trailing digit needed
+                    1;
+
+            case 0, 1, 2, 3, 4, 5, 6 ->
+                // Keep digits to left of decimal, plus leave a
+                // trailing zero
+                    (expR + 1) + 1;
+
+            default ->
+                // Require at least 2 digits, to include trailing digit when
+                // there is a single digit
+                    2;
+        };
+
+        long s = fd.f;
+        int prec = fd.n;
+
+        if (prec < targetPrec) {
+            // Add zeros needed to reach target precision
+            int addZeros = targetPrec - prec;
+            s *= MathUtils.pow10(addZeros); // addZeros will be at most 8
+            prec = targetPrec;
+        } else {
+            // Remove trailing zeros to try to reach target precision
+            while (prec > targetPrec && s % 10 == 0) {
+                s = s / 10;
+                prec--;
+            }
+        }
+
+        // Calculate new e based on updated precision
+        int eNew = expR - prec + 1;  // expR is defined as prec + e - 1
+        fd.set(s, eNew, prec);
+
+        return fd;
+    }
+
+    public long getSignificand() {
+        return f;
+    }
+
+    public int getPrecision() {
+        return n;
+    }
+
+    /**
+     * Value {@code scale}, such that
+     * {@code value = f * 10^(-scale)}
+     * <p/>
+     * This is the same as the definition used by {@link java.math.BigDecimal}
+     */
+    public int getScale() {
+        return -e;
+    }
```

Comment:
I'd prefer to have a `getExp()` rather than `getScale()` here, as everything here uses an exponent rather than a scale.
Then there's no need for the JavaDoc comment.